088-rebalance-progress-status

```diff
@@ -0,0 +1,328 @@
+
+All information required for estimating the values of  “estimatedTimeToCompletion” and  “percentageComplete” fields can be derived from either Cruise Control server configurations or CruiseControlState endpoint. 
+That being said, the method of estimation for these fields depends on the state of the KafkaRebalance resource.
+
+#### estimatedTimeToCompletion
+
+##### Rebalancing
+
+```
+rate = (finishedDataMovement)/(<task_trigger_time> - <current_time>)
+
+estimatedTimeToCompletion = (totalDateToMove-finishedDataMovement) / (rate)
+```
+
+##### Stopped
+
+Once a rebalance has been stopped, it cannot be completed. 
+Therefore, there is no “estimationTimeToCompletion” for a stopped rebalance, so we set estimatedTimeToCompletion = null to emphasize this. 
+The `KafkaRebalance` resource must be refreshed and the progress section overwritten with the next state change. 
+
+```
+estimatedTimeToCompletion = null
+```
+
+
+#### percentageComplete
+
+##### Rebalancing
+
+```
+percentageComplete = (finishedDataMovement/totalDataToMoveMB)%
+```
+
+##### Stopped
+
+Once a rebalance has been stopped, it cannot be completed. The `KafkaRebalance` resource must be refreshed and the progress section overwritten with that change. That being said, before the `KafkaRebalace` resource is deleted or “refreshed”, the percentageComplete information will still be of value to users. 
+
+```
+percentageComplete = (finishedDataMovement/totalDataToMoveMB)
+```
+
+#### rebalanceProgressConfigMap
+
+Will only be present in “Rebalancing” and “Stopped” states.
+
+The enhanced `KafkaRebalance` resource would include the following in its status section
+
+```
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+spec: {}
+status:
+  conditions:
+  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
+	status: "True"
+	type: Rebalancing | Stopped [1]
+  observedGeneration: 1
+  optimizationResult:
+	afterBeforeLoadConfigMap: my-rebalance
+	dataToMoveMB: 0
+	excludedBrokersForLeadership: []
+	excludedBrokersForReplicaMove: []
+	excludedTopics: []
+	intraBrokerDataToMoveMB: 0
+	monitoredPartitionsPercentage: 100
+	numIntraBrokerReplicaMovements: 0
+	numLeaderMovements: 16
+	numReplicaMovements: 0
+	onDemandBalancednessScoreAfter: 95.4347095948149
+	onDemandBalancednessScoreBefore: 89.4347095948149
+	provisionRecommendation: ""
+	provisionStatus: RIGHT_SIZED
+	recentWindows: 1
+  progress:
+        estimatedTimeToCompletion: 5m [2]
+        percentageComplete: 80% [3]
+        rebalanceProgressConfigMap: my-rebalance-progress [4]
+```
+[1] The “progress” section will be visible during the KafkaRebalance “Rebalancing” and  “Stopped” states.
+[2] The minimum estimated time it will take the rebalance to complete.
+[3] The percentage complete of the ongoing rebalance in the range [0-100]%
+[4] The ConfigMap where “non-verbose” JSON payload from Executor State from CruiseControlState endpoint is stored.
```

Comment:
This is the first time you have mentioned the payload. I would either talk about this in the config map section above or link/mention the executor state section that explains what this is.

Reply:
Why do we need to store the state in a config map? Maybe we could simply document how to recover that from the REST endpoint in case it is needed for troubleshooting.

Reply:
@fvaleri which kind of troubleshooting are you talking about? The verbose information stored in the ConfigMap are coming from the `state?substates=executor` endpoint which has data when something is running, otherwise it just returns a `NO_TASK_IN_PROGRESS` so in case of issues, you can't get anything interesting from here AFAIK.